src/umode.rs

```diff
@@ -356,6 +372,14 @@ impl UmodeTask {
                     ControlFlow::Continue(())
                 }
                 HypCall::NextOp(result) => ControlFlow::Break(Ok(result)),
+                hypc => {
+                    // Handle the rest through the caller-supplied `hypc_handler`.
+                    // All errors will immediately stop and reset U-mode.
+                    match hypc_handler(hypc) {
+                        Ok(()) => ControlFlow::Continue(()),
+                        Err(e) => ControlFlow::Break(Err(e)),
+                    }
+                }
```

Comment:
The intention of this is to only allow certain hypercalls for certain requests, right? e.g. only allowing signing for requests that have a need for it. Instead of a clousre, could you have a `UmodeRequest` method that takes a `HypCall` saying whether or not that `HypCall` is allowed? (or vice-versa)

Reply:
I also have a feeling that a function as a handler to filter what is allowed might not be the best way to manage the set of available hypervisor calls. I think you'll probably want something that's easy to compose (in many cases, you'll probably want the standard hypercalls plus a small number of additional calls). Perhaps we can learn from similar problems elsewhere, and do something similar to a syscall table here? In its simplest form it's just a map of HypCall to optional handler function, and it'd be simple enough to compose (start with the standard handler table, add in the signing one).

Now, you since a handler function is maximally generic, you can do all I say above with what you have, I guess I'm just wondering how this will look like in practice when we have more than one  umode request with more than one special-purpose restricted hypercalls...

(As often with my comments, just some unsolicited thinking-ahead, please feel free to ignore/defer until it becomes more relevant)

Reply:
UmodeRequest is meant for the API and not for the machinery of the U-mode loop. So that will require a differfent type.

Having said this, I strongly prefer to use a closure because it will define data specific to the call. For example, which key to sign, how to get the key, etc., without requiring the umode module to be aware of all the internal specifics of complex requests.

We could be a bit more precise and define a "External Hypcall" type, so to better specify that this closure only specifies extra hypercalls, and doesn't filter the standard ones.

This is not filtering, this is defining some extra hypercalls that are more complex than simple management of the U-mode execution state.